File: test/TestCases/node-addon-api/basic_types/value.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using Microsoft.JavaScript.NodeApi;

namespace Microsoft.JavaScript.NodeApiTest;

public class TestBasicTypesValue : TestHelper, ITestObject
{
    private static JSValue IsUndefined(JSCallbackArgs args) => args[0].IsUndefined();
    private static JSValue IsNull(JSCallbackArgs args) => args[0].IsNull();
    private static JSValue IsBoolean(JSCallbackArgs args) => args[0].IsBoolean();
    private static JSValue IsNumber(JSCallbackArgs args) => args[0].IsNumber();
    private static JSValue IsString(JSCallbackArgs args) => args[0].IsString();
    private static JSValue IsSymbol(JSCallbackArgs args) => args[0].IsSymbol();
    private static JSValue IsArray(JSCallbackArgs args) => args[0].IsArray();
    private static JSValue IsArrayBuffer(JSCallbackArgs args) => args[0].IsArrayBuffer();
    private static JSValue IsTypedArray(JSCallbackArgs args) => args[0].IsTypedArray();
    private static JSValue IsObject(JSCallbackArgs args) => args[0].IsObject();
    private static JSValue IsFunction(JSCallbackArgs args) => args[0].IsFunction();
    private static JSValue IsPromise(JSCallbackArgs args) => args[0].IsPromise();
    private static JSValue IsDataView(JSCallbackArgs args) => args[0].IsDataView();
    private static JSValue IsExternal(JSCallbackArgs args) => args[0].IsExternal();
    private static JSValue ToBoolean(JSCallbackArgs args) => args[0].CoerceToBoolean();
    private static JSValue ToNumber(JSCallbackArgs args) => args[0].CoerceToNumber();
    private static JSValue ToString(JSCallbackArgs args) => args[0].CoerceToString();
    private static JSValue ToObject(JSCallbackArgs args) => args[0].CoerceToObject();

    private static JSValue StrictlyEquals(JSCallbackArgs args) => args[0].Equals(args[1]);

    // Helper methods
    private static JSValue CreateDefaultValue(JSCallbackArgs _) => default;
    private static JSValue CreateEmptyValue(JSCallbackArgs _) => new();
    private static JSValue CreateNonEmptyValue(JSCallbackArgs _) => "non_empty_val";
    private static JSValue CreateExternal(JSCallbackArgs _) => JSValue.CreateExternal(1);


    public JSObject Init() => new()
    {
        Method(IsUndefined, nameof(IsUndefined)),
        Method(IsNull, nameof(IsNull)),
        Method(IsBoolean, nameof(IsBoolean)),
        Method(IsNumber, nameof(IsNumber)),
        Method(IsString, nameof(IsString)),
        Method(IsSymbol, nameof(IsSymbol)),
        Method(IsArray, nameof(IsArray)),
        Method(IsArrayBuffer, nameof(IsArrayBuffer)),
        Method(IsTypedArray, nameof(IsTypedArray)),
        Method(IsObject, nameof(IsObject)),
        Method(IsFunction, nameof(IsFunction)),
        Method(IsPromise, nameof(IsPromise)),
        Method(IsDataView, nameof(IsDataView)),
        Method(IsExternal, nameof(IsExternal)),
        Method(ToBoolean, nameof(ToBoolean)),
        Method(ToNumber, nameof(ToNumber)),
        Method(ToString, nameof(ToString)),
        Method(ToObject, nameof(ToObject)),

        Method(StrictlyEquals, nameof(StrictlyEquals)),

        Method(CreateDefaultValue, nameof(CreateDefaultValue)),
        Method(CreateEmptyValue, nameof(CreateEmptyValue)),
        Method(CreateNonEmptyValue, nameof(CreateNonEmptyValue)),
        Method(CreateExternal, nameof(CreateExternal)),
    };
}

File: test/TestCases/node-addon-api/basic_types/value.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

'use strict';

const assert = require('assert');

module.exports = require('../../common').runTest(test);

function test(binding) {
  const externalValue = binding.basicTypesValue.createExternal();

  function isObject(value) {
    return (typeof value === 'object' && value !== externalValue) ||
      (typeof value === 'function');
  }

  function detailedTypeOf(value) {
    const type = typeof value;
    if (type !== 'object') { return type; }

    if (value === null) { return 'null'; }

    if (Array.isArray(value)) { return 'array'; }

    if (value === externalValue) { return 'external'; }

    if (!value.constructor) { return type; }

    if (value instanceof ArrayBuffer) { return 'arraybuffer'; }

    if (ArrayBuffer.isView(value)) {
      if (value instanceof DataView) {
        return 'dataview';
      } else {
        return 'typedarray';
      }
    }

    if (value instanceof Promise) { return 'promise'; }

    return 'object';
  }

  function typeCheckerTest(typeChecker, expectedType) {
    const testValueList = [
      undefined,
      null,
      true,
      10,
      'string',
      Symbol('symbol'),
      [],
      new ArrayBuffer(10),
      new Int32Array(new ArrayBuffer(12)),
      {},
      function () { },
      new Promise((resolve, reject) => { }),
      new DataView(new ArrayBuffer(12)),
      externalValue
    ];

    testValueList.forEach((testValue, index) => {
      if (testValue !== null && expectedType === 'object') {
        assert.strictEqual(typeChecker(testValue), isObject(testValue), `Index: ${index}`);
      } else {
        assert.strictEqual(typeChecker(testValue), detailedTypeOf(testValue) === expectedType, `Index: ${index}`);
      }
    });
  }

  function typeConverterTest(typeConverter, expectedType) {
    const testValueList = [
      true,
      false,
      0,
      10,
      'string',
      [],
      new ArrayBuffer(10),
      new Int32Array(new ArrayBuffer(12)),
      {},
      function () { },
      new Promise((resolve, reject) => { })
    ];

    testValueList.forEach((testValue, index) => {
      const expected = expectedType(testValue);
      const actual = typeConverter(testValue);

      if (isNaN(expected)) {
        assert(isNaN(actual), `Index: ${index}`);
      } else {
        assert.deepStrictEqual(typeConverter(testValue), expectedType(testValue), `Index: ${index}`);
      }
    });
  }

  function assertValueStrictlyEqual(value) {
    const newValue = value.createNonEmptyValue();
    assert(value.strictlyEquals(newValue, newValue));
  }

  function assertValueStrictlyNonEqual(value) {
    const valueA = value.createNonEmptyValue();
    const valueB = value.createExternal();
    assert(!value.strictlyEquals(valueA, valueB));
  }

   const value = binding.basicTypesValue;

  assertValueStrictlyEqual(value);
  assertValueStrictlyNonEqual(value);

  typeCheckerTest(value.isUndefined, 'undefined');
  typeCheckerTest(value.isNull, 'null');
  typeCheckerTest(value.isBoolean, 'boolean');
  typeCheckerTest(value.isNumber, 'number');
  typeCheckerTest(value.isString, 'string');
  typeCheckerTest(value.isSymbol, 'symbol');
  typeCheckerTest(value.isArray, 'array');
  typeCheckerTest(value.isArrayBuffer, 'arraybuffer');
  typeCheckerTest(value.isTypedArray, 'typedarray');
  typeCheckerTest(value.isObject, 'object');
  typeCheckerTest(value.isFunction, 'function');
  typeCheckerTest(value.isPromise, 'promise');
  typeCheckerTest(value.isDataView, 'dataview');
  typeCheckerTest(value.isExternal, 'external');

  assert.strictEqual(value.isUndefined(value.createDefaultValue()), true);
  assert.strictEqual(value.isUndefined(value.createEmptyValue()), true);

  typeConverterTest(value.toBoolean, Boolean);
  assert.strictEqual(value.toBoolean(undefined), false);
  assert.strictEqual(value.toBoolean(null), false);

  typeConverterTest(value.toNumber, Number);
  assert(isNaN(value.toNumber(undefined)));
  assert.strictEqual(value.toNumber(null), 0);

  typeConverterTest(value.toString, String);
  assert.strictEqual(value.toString(undefined), 'undefined');
  assert.strictEqual(value.toString(null), 'null');

  typeConverterTest(value.toObject, Object);
}
